enhancements/microshift/microshift-supportability-tools.md

```diff
@@ -0,0 +1,384 @@
+---
+title: MicroShift supportability tools
+authors:
+  - pacevedom
+reviewers:
+  - "@copejon, MicroShift contributor"
+  - "@fzdarsky, MicroShift architect"
+  - "@ggiguash, MicroShift contributor"
+  - "@oglok, MicroShift contributor"
+  - "@pmtk, MicroShift contributor"
+  - "@TurboTurtle, SOS contributor"
+approvers:
+  - "@dhellmann"
+api-approvers:
+  - None
+creation-date: 2023-01-17
+last-updated: 2023-01-17
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-389
+see-also:
+  - "/enhancements/oc/must-gather.md"
+  - "https://sos.readthedocs.org/"
+---
+
+# MicroShift supportability tools
+
+## Summary
+This enhancement proposes the addition of supportability tooling to MicroShift.
+
+To debug issues in the cluster we need a single command that any user or
+administrator is able to run to gather all the required information to diagnose
+and solve problems.
+
+## Motivation
+Most readers will be familiar with `must-gather` for OpenShift clusters and
+what it achieves. For MicroShift we need a tool with the same principles and
+goals, which is to gather every bit of information from the cluster to be able
+to diagnose and fix issues.
+
+### User Stories
+
+* As a MicroShift administrator, I want to be able to get a report about current
+  cluster status.
+* As a MicroShift developer, I want to be able to diagnose issues with a detailed
+  debug report.
+
+### Goals
+* Provide a single tool to automate the report creation.
+
+### Non-Goals
+N/A
+
+## Proposal
+The MicroShift gathering tool is a plugin in `sos`. This tool is delivered
+with RHEL, and is capable of generating reports for many applications in the
+system.
+
+Executing `sos report` yields information about a system and its components by
+using plugins to read each one of them. Plugins have an enablement check or may
+be invoked manually when executing the command.
+A [plugin](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin) is a
+python script that we include as part of `sos`.
+
+There is a [plugin for OpenShift](https://github.com/sosreport/sos/blob/main/sos/report/plugins/openshift.py)
+already that almost fits the needs of this EP. MicroShift does not share the
+same resources, APIs and namespaces/names that the plugin assumes, as it was
+made for OpenShift clusters. However, much of the needs covered in the plugin
+are the same, so the contents of the plugin may be used too.
```

Comment:
Almost none (unless you count "Red Hat Plugin X" and "Ubuntu Plugin X" as different plugins), but that's not for any particular code level reason. It's just a side effect of the fact that plugins are component/package specific, but I/we have no fundamental objection to plugin reuse.

I have a draft microshift plugin drawn up that reuses the openshift plugin, but sets a different enablement trigger and ignores the oauth bits since `oc login` doesn't work on microshift (or so I'm told). It'd be very simple to further make it so that the openshift and microshift plugins define their own list of resources and/or namespaces to collect from, but the collection logic remains the same (e.g. `oc get $resource` then `oc describe $resource $specific_entity` for everything reported by `oc get`).

If you like, I can post a WIP to sos for this plugin in a day or so for review and we can refine from there.

Reply:
Nice, I really expected we would end up needing to write that. It would be great for you and @pacevedom to work together on it, as well as this design doc.

Reply:
Ack - draft is up here for reference - https://github.com/sosreport/sos/pull/3116

Reply:
Thanks @TurboTurtle! Let me use this as a base for refining the end result. Will post up changes when done.

Reply:
Taking this on sos repo.